x/evm/keeper/state_transition.go

```diff
@@ -231,20 +232,22 @@ func (k *Keeper) ApplyMessage(evm *vm.EVM, msg core.Message, cfg *params.ChainCo
 		ret, leftoverGas, vmErr = evm.Call(sender, *msg.To(), msg.Data(), leftoverGas, msg.Value())
 	}
 
-	var reverted bool
-	if vmErr != nil {
-		if !errors.Is(vmErr, vm.ErrExecutionReverted) {
-			// wrap the VM error
-			return nil, stacktrace.Propagate(sdkerrors.Wrap(types.ErrVMExecution, vmErr.Error()), "vm execution failed")
+	if simulate {
+		// eth_call/eth_estimateGas don't call ante handler to deduct gas fee, so don't do actual refund.
+		leftoverGas += k.GasToRefund(msg.Gas() - leftoverGas)
+	} else {
+		// refund gas prior to handling the vm error in order to set the updated gas meter
+		leftoverGas, err = k.RefundGas(msg, leftoverGas)
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "failed to refund gas leftover gas to sender %s", msg.From())
 		}
-		reverted = true
 	}
 
 	gasUsed := msg.Gas() - leftoverGas
-	return &types.MsgEthereumTxResponse{
-		Ret:      ret,
-		Reverted: reverted,
-		GasUsed:  gasUsed,
+	return &core.ExecutionResult{
```

Comment:
are we using the `ExecutionResult` from go-ethereum just to make it easier to check for failures? can we add a comment on this decision too 🙏

Reply:
Changed `ExecuteResult` back to `MsgEthereumTxResponse`, since `MsgEthereumTxResponse` contains all we need now.